### ISSUE TITLE
Test Errors and Failures while trying to Package for openSUSE

### ISSUE THREAD

**Issue by Pi-Cla** [qtile/qtile#3682]
### The issue:

The Qtile package for openSUSE has not been updated in a while and is currently failing some tests, (though better than Fedora where the package no longer exists I guess...) so I decided to actually try to update it and fix the failing tests. I had to use the newer ffibuild script 8f70047ea5c24f484e527ea237e5b72c1ba5ff47 as openSUSE does not actually have pywayland or pywlroots packaged yet, and I also had to use the patch at #3518 to fix some of the build failures. (I also decided to add #3544 as it didn't seem to impact the tests and looks like a nice feature)

With the preface out of the way, here is the main issue now. Even after applying the above patches, there are still a bunch of test failures. The full log file can be found at: https://build.opensuse.org/package/live_build_log/home:Pi-Cla:branches:X11:windowmanagers/qtile/openSUSE_Tumbleweed/x86_64 but the errors get duplicated a lot across tests so I will just list them once. Also, you can see which patches I am currently using at https://build.opensuse.org/package/show/home:Pi-Cla:branches:X11:windowmanagers/qtile.

```python
[ 2221s] _____________ ERROR at setup of test_basic[1-x11-FakeScreenConfig] _____________
[ 2221s] 
[ 2221s] request = <SubRequest 'manager' for <Function test_basic[1-x11-FakeScreenConfig]>>
[ 2221s] manager_nospawn = <test.helpers.TestManager object at 0x7f92c8fd6170>
[ 2221s] 
[ 2221s]     @pytest.fixture(scope="function")
[ 2221s]     def manager(request, manager_nospawn):
[ 2221s]         config = getattr(request, "param", BareConfig)
[ 2221s]     
[ 2221s] >       manager_nospawn.start(config)
[ 2221s] 
[ 2221s] test/conftest.py:139: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] self = <test.helpers.TestManager object at 0x7f92c8fd6170>
[ 2221s] config_class = <class 'test.test_fakescreen.FakeScreenConfig'>, no_spawn = False
[ 2221s] state = None
[ 2221s] 
[ 2221s]     def start(self, config_class, no_spawn=False, state=None):
[ 2221s]         rpipe, wpipe = multiprocessing.Pipe()
[ 2221s]     
[ 2221s]         def run_qtile():
[ 2221s]             try:
[ 2221s]                 os.environ.pop("DISPLAY", None)
[ 2221s]                 os.environ.[ 2221s] _____________ ERROR at setup of test_basic[1-x11-FakeScreenConfig] _____________
[ 2221s] 
[ 2221s] request = <SubRequest 'manager' for <Function test_basic[1-x11-FakeScreenConfig]>>
[ 2221s] manager_nospawn = <test.helpers.TestManager object at 0x7f92c8fd6170>
[ 2221s] 
[ 2221s]     @pytest.fixture(scope="function")
[ 2221s]     def manager(request, manager_nospawn):
[ 2221s]         config = getattr(request, "param", BareConfig)
[ 2221s]     
[ 2221s] >       manager_nospawn.start(config)
[ 2221s] 
[ 2221s] test/conftest.py:139: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] self = <test.helpers.TestManager object at 0x7f92c8fd6170>
[ 2221s] config_class = <class 'test.test_fakescreen.FakeScreenConfig'>, no_spawn = False
[ 2221s] state = None
[ 2221s] 
[ 2221s]     def start(self, config_class, no_spawn=False, state=None):
[ 2221s]         rpipe, wpipe = multiprocessing.Pipe()
[ 2221s]     
[ 2221s]         def run_qtile():
[ 2221s]             try:
[ 2221s]                 os.environ.pop("DISPLAY", None)
[ 2221s]                 os.environ.pop("WAYLAND_DISPLAY", None)
[ 2221s]                 kore = self.backend.create()
[ 2221s]                 os.environ.update(self.backend.env)
[ 2221s]                 logger = init_log(self.log_level, log_path=None, log_color=False)
[ 2221s]                 if hasattr(self, "log_queue"):
[ 2221s]                     logger.addHandler(logging.handlers.QueueHandler(self.log_queue))
[ 2221s]                 Qtile(
[ 2221s]                     kore,
[ 2221s]                     config_class(),
[ 2221s]                     socket_path=self.sockfile,
[ 2221s]                     no_spawn=no_spawn,
[ 2221s]                     state=state,
[ 2221s]                 ).loop()
[ 2221s]             except Exception:
[ 2221s]                 wpipe.send(traceback.format_exc())
[ 2221s]     
[ 2221s]         self.proc = multiprocessing.Process(target=run_qtile)
[ 2221s]         self.proc.start()
[ 2221s]     
[ 2221s]         # First, wait for socket to appear
[ 2221s]         if can_connect_qtile(self.sockfile, ok=lambda: not rpipe.poll()):
[ 2221s]             ipc_client = ipc.Client(self.sockfile)
[ 2221s]             ipc_command = command.interface.IPCCommandInterface(ipc_client)
[ 2221s]             self.c = command.client.InteractiveCommandClient(ipc_command)
[ 2221s]             self.backend.configure(self)
[ 2221s]             return
[ 2221s]         if rpipe.poll(0.1):
[ 2221s]             error = rpipe.recv()
[ 2221s]             raise AssertionError("Error launching qtile, traceback:\n%s" % error)
[ 2221s] >       raise AssertionError("Error launching qtile")
[ 2221s] E       AssertionError: Error launching qtile
[ 2221s] 
[ 2221s] test/helpers.py:203: AssertionError
[ 2221s] ---------------------------- Captured stdout setup -----------------------------
[ 2221s] 2022-07-07 07:53:56,726 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: sep_1, importerrorwidget_1, importerrorwidget_2, sep_2 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,839 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_1, windowname_1, clock_1 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,843 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_2, windowname_2, clock_2 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,847 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_3, windowname_3, clock_3 To bind commands, rename the widget or use lazy.widget[new_name].pop("WAYLAND_DISPLAY", None)
[ 2221s]                 kore = self.backend.create()
[ 2221s]                 os.environ.update(self.backend.env)
[ 2221s]                 logger = init_log(self.log_level, log_path=None, log_color=False)
[ 2221s]                 if hasattr(self, "log_queue"):
[ 2221s]                     logger.addHandler(logging.handlers.QueueHandler(self.log_queue))
[ 2221s]                 Qtile(
[ 2221s]                     kore,
[ 2221s]                     config_class(),
[ 2221s]                     socket_path=self.sockfile,
[ 2221s]                     no_spawn=no_spawn,
[ 2221s]                     state=state,
[ 2221s]                 ).loop()
[ 2221s]             except Exception:
[ 2221s]                 wpipe.send(traceback.format_exc())
[ 2221s]     
[ 2221s]         self.proc = multiprocessing.Process(target=run_qtile)
[ 2221s]         self.proc.start()
[ 2221s]     
[ 2221s]         # First, wait for socket to appear
[ 2221s]         if can_connect_qtile(self.sockfile, ok=lambda: not rpipe.poll()):
[ 2221s]             ipc_client = ipc.Client(self.sockfile)
[ 2221s]             ipc_command = command.interface.IPCCommandInterface(ipc_client)
[ 2221s]             self.c = command.client.InteractiveCommandClient(ipc_command)
[ 2221s]             self.backend.configure(self)
[ 2221s]             return
[ 2221s]         if rpipe.poll(0.1):
[ 2221s]             error = rpipe.recv()
[ 2221s]             raise AssertionError("Error launching qtile, traceback:\n%s" % error)
[ 2221s] >       raise AssertionError("Error launching qtile")
[ 2221s] E       AssertionError: Error launching qtile
[ 2221s] 
[ 2221s] test/helpers.py:203: AssertionError
[ 2221s] ---------------------------- Captured stdout setup -----------------------------
[ 2221s] 2022-07-07 07:53:56,726 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: sep_1, importerrorwidget_1, importerrorwidget_2, sep_2 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,839 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_1, windowname_1, clock_1 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,843 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_2, windowname_2, clock_2 To bind commands, rename the widget or use lazy.widget[new_name].
[ 2221s] 2022-07-07 07:53:56,847 INFO libqtile bar.py:_configure():L315 The following widgets were renamed in qtile.widgets_map: groupbox_3, windowname_3, clock_3 To bind commands, rename the widget or use lazy.widget[new_name].
```
```python
[ 2221s] _____________________ test_cycle_layouts[1-x11-AllLayouts] _____________________
[ 2221s] 
[ 2221s] self = <StreamReader eof transport=<_SelectorSocketTransport closed fd=56>>
[ 2221s] n = -1
[ 2221s] 
[ 2221s]     async def read(self, n=-1):
[ 2221s]         """Read up to `n` bytes from the stream.
[ 2221s]     
[ 2221s]         If n is not provided, or set to -1, read until EOF and return all read
[ 2221s]         bytes. If the EOF was received and the internal buffer is empty, return
[ 2221s]         an empty bytes object.
[ 2221s]     
[ 2221s]         If n is zero, return empty bytes object immediately.
[ 2221s]     
[ 2221s]         If n is positive, this function try to read `n` bytes, and may return
[ 2221s]         less or equal bytes than requested, but at least one byte. If EOF was
[ 2221s]         received before any byte is read, this function returns empty byte
[ 2221s]         object.
[ 2221s]     
[ 2221s]         Returned value is not limited with limit, configured at stream
[ 2221s]         creation.
[ 2221s]     
[ 2221s]         If stream was paused, this function will automatically resume it if
[ 2221s]         needed.
[ 2221s]         """
[ 2221s]     
[ 2221s]         if self._exception is not None:
[ 2221s]             raise self._exception
[ 2221s]     
[ 2221s]         if n == 0:
[ 2221s]             return b''
[ 2221s]     
[ 2221s]         if n < 0:
[ 2221s]             # This used to just loop creating a new waiter hoping to
[ 2221s]             # collect everything in self._buffer, but that would
[ 2221s]             # deadlock if the subprocess sends more than self.limit
[ 2221s]             # bytes.  So just call self.read(self._limit) until EOF.
[ 2221s]             blocks = []
[ 2221s]             while True:
[ 2221s] >               block = await self.read(self._limit)
[ 2221s] 
[ 2221s] /usr/lib64/python3.10/asyncio/streams.py:662: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] self = <StreamReader eof transport=<_SelectorSocketTransport closed fd=56>>
[ 2221s] n = 65536
[ 2221s] 
[ 2221s]     async def read(self, n=-1):
[ 2221s]         """Read up to `n` bytes from the stream.
[ 2221s]     
[ 2221s]         If n is not provided, or set to -1, read until EOF and return all read
[ 2221s]         bytes. If the EOF was received and the internal buffer is empty, return
[ 2221s]         an empty bytes object.
[ 2221s]     
[ 2221s]         If n is zero, return empty bytes object immediately.
[ 2221s]     
[ 2221s]         If n is positive, this function try to read `n` bytes, and may return
[ 2221s]         less or equal bytes than requested, but at least one byte. If EOF was
[ 2221s]         received before any byte is read, this function returns empty byte
[ 2221s]         object.
[ 2221s]     
[ 2221s]         Returned value is not limited with limit, configured at stream
[ 2221s]         creation.
[ 2221s]     
[ 2221s]         If stream was paused, this function will automatically resume it if
[ 2221s]         needed.
[ 2221s]         """
[ 2221s]     
[ 2221s]         if self._exception is not None:
[ 2221s]             raise self._exception
[ 2221s]     
[ 2221s]         if n == 0:
[ 2221s]             return b''
[ 2221s]     
[ 2221s]         if n < 0:
[ 2221s]             # This used to just loop creating a new waiter hoping to
[ 2221s]             # collect everything in self._buffer, but that would
[ 2221s]             # deadlock if the subprocess sends more than self.limit
[ 2221s]             # bytes.  So just call self.read(self._limit) until EOF.
[ 2221s]             blocks = []
[ 2221s]             while True:
[ 2221s]                 block = await self.read(self._limit)
[ 2221s]                 if not block:
[ 2221s]                     break
[ 2221s]                 blocks.append(block)
[ 2221s]             return b''.join(blocks)
[ 2221s]     
[ 2221s]         if not self._buffer and not self._eof:
[ 2221s] >           await self._wait_for_data('read')
[ 2221s] 
[ 2221s] /usr/lib64/python3.10/asyncio/streams.py:669: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] self = <StreamReader eof transport=<_SelectorSocketTransport closed fd=56>>
[ 2221s] func_name = 'read'
[ 2221s] 
[ 2221s]     async def _wait_for_data(self, func_name):
[ 2221s]         """Wait until feed_data() or feed_eof() is called.
[ 2221s]     
[ 2221s]         If stream was paused, automatically resume it.
[ 2221s]         """
[ 2221s]         # StreamReader uses a future to link the protocol feed_data() method
[ 2221s]         # to a read coroutine. Running two read coroutines at the same time
[ 2221s]         # would have an unexpected behaviour. It would not possible to know
[ 2221s]         # which coroutine would get the next data.
[ 2221s]         if self._waiter is not None:
[ 2221s]             raise RuntimeError(
[ 2221s]                 f'{func_name}() called while another coroutine is '
[ 2221s]                 f'already waiting for incoming data')
[ 2221s]     
[ 2221s]         assert not self._eof, '_wait_for_data after EOF'
[ 2221s]     
[ 2221s]         # Waiting for data while paused will make deadlock, so prevent it.
[ 2221s]         # This is essential for readexactly(n) for case when n > self._limit.
[ 2221s]         if self._paused:
[ 2221s]             self._paused = False
[ 2221s]             self._transport.resume_reading()
[ 2221s]     
[ 2221s]         self._waiter = self._loop.create_future()
[ 2221s]         try:
[ 2221s] >           await self._waiter
[ 2221s] E           asyncio.exceptions.CancelledError
[ 2221s] 
[ 2221s] /usr/lib64/python3.10/asyncio/streams.py:502: CancelledError
[ 2221s] 
[ 2221s] During handling of the above exception, another exception occurred:
[ 2221s] 
[ 2221s] fut = <Task cancelled name='Task-851273' coro=<StreamReader.read() done, defined at /usr/lib64/python3.10/asyncio/streams.py:628>>
[ 2221s] timeout = 10
[ 2221s] 
[ 2221s]     async def wait_for(fut, timeout):
[ 2221s]         """Wait for the single Future or coroutine to complete, with timeout.
[ 2221s]     
[ 2221s]         Coroutine will be wrapped in Task.
[ 2221s]     
[ 2221s]         Returns result of the Future or coroutine.  When a timeout occurs,
[ 2221s]         it cancels the task and raises TimeoutError.  To avoid the task
[ 2221s]         cancellation, wrap it in shield().
[ 2221s]     
[ 2221s]         If the wait is cancelled, the task is also cancelled.
[ 2221s]     
[ 2221s]         This function is a coroutine.
[ 2221s]         """
[ 2221s]         loop = events.get_running_loop()
[ 2221s]     
[ 2221s]         if timeout is None:
[ 2221s]             return await fut
[ 2221s]     
[ 2221s]         if timeout <= 0:
[ 2221s]             fut = ensure_future(fut, loop=loop)
[ 2221s]     
[ 2221s]             if fut.done():
[ 2221s]                 return fut.result()
[ 2221s]     
[ 2221s]             await _cancel_and_wait(fut, loop=loop)
[ 2221s]             try:
[ 2221s]                 return fut.result()
[ 2221s]             except exceptions.CancelledError as exc:
[ 2221s]                 raise exceptions.TimeoutError() from exc
[ 2221s]     
[ 2221s]         waiter = loop.create_future()
[ 2221s]         timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
[ 2221s]         cb = functools.partial(_release_waiter, waiter)
[ 2221s]     
[ 2221s]         fut = ensure_future(fut, loop=loop)
[ 2221s]         fut.add_done_callback(cb)
[ 2221s]     
[ 2221s]         try:
[ 2221s]             # wait until the future completes or the timeout
[ 2221s]             try:
[ 2221s]                 await waiter
[ 2221s]             except exceptions.CancelledError:
[ 2221s]                 if fut.done():
[ 2221s]                     return fut.result()
[ 2221s]                 else:
[ 2221s]                     fut.remove_done_callback(cb)
[ 2221s]                     # We must ensure that the task is not running
[ 2221s]                     # after wait_for() returns.
[ 2221s]                     # See https://bugs.python.org/issue32751
[ 2221s]                     await _cancel_and_wait(fut, loop=loop)
[ 2221s]                     raise
[ 2221s]     
[ 2221s]             if fut.done():
[ 2221s]                 return fut.result()
[ 2221s]             else:
[ 2221s]                 fut.remove_done_callback(cb)
[ 2221s]                 # We must ensure that the task is not running
[ 2221s]                 # after wait_for() returns.
[ 2221s]                 # See https://bugs.python.org/issue32751
[ 2221s]                 await _cancel_and_wait(fut, loop=loop)
[ 2221s]                 # In case task cancellation failed with some
[ 2221s]                 # exception, we should re-raise it
[ 2221s]                 # See https://bugs.python.org/issue40607
[ 2221s]                 try:
[ 2221s] >                   return fut.result()
[ 2221s] E                   asyncio.exceptions.CancelledError
[ 2221s] 
[ 2221s] /usr/lib64/python3.10/asyncio/tasks.py:456: CancelledError
[ 2221s] 
[ 2221s] The above exception was the direct cause of the following exception:
[ 2221s] 
[ 2221s] self = <libqtile.ipc.Client object at 0x7f92c8b275b0>
[ 2221s] msg = ([], 'next_layout', (), {})
[ 2221s] 
[ 2221s]     async def async_send(self, msg: Any) -> Any:
[ 2221s]         """Send the message to the server
[ 2221s]     
[ 2221s]         Connect to the server, then pack and send the message to the server,
[ 2221s]         then wait for and return the response from the server.
[ 2221s]         """
[ 2221s]         try:
[ 2221s]             reader, writer = await asyncio.wait_for(
[ 2221s]                 asyncio.open_unix_connection(path=self.socket_path), timeout=3
[ 2221s]             )
[ 2221s]         except (ConnectionRefusedError, FileNotFoundError):
[ 2221s]             raise IPCError("Could not open {}".format(self.socket_path))
[ 2221s]     
[ 2221s]         try:
[ 2221s]             send_data = _IPC.pack(msg, is_json=self.is_json)
[ 2221s]             writer.write(send_data)
[ 2221s]             writer.write_eof()
[ 2221s]     
[ 2221s] >           read_data = await asyncio.wait_for(reader.read(), timeout=10)
[ 2221s] 
[ 2221s] libqtile/ipc.py:191: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] fut = <Task cancelled name='Task-851273' coro=<StreamReader.read() done, defined at /usr/lib64/python3.10/asyncio/streams.py:628>>
[ 2221s] timeout = 10
[ 2221s] 
[ 2221s]     async def wait_for(fut, timeout):
[ 2221s]         """Wait for the single Future or coroutine to complete, with timeout.
[ 2221s]     
[ 2221s]         Coroutine will be wrapped in Task.
[ 2221s]     
[ 2221s]         Returns result of the Future or coroutine.  When a timeout occurs,
[ 2221s]         it cancels the task and raises TimeoutError.  To avoid the task
[ 2221s]         cancellation, wrap it in shield().
[ 2221s]     
[ 2221s]         If the wait is cancelled, the task is also cancelled.
[ 2221s]     
[ 2221s]         This function is a coroutine.
[ 2221s]         """
[ 2221s]         loop = events.get_running_loop()
[ 2221s]     
[ 2221s]         if timeout is None:
[ 2221s]             return await fut
[ 2221s]     
[ 2221s]         if timeout <= 0:
[ 2221s]             fut = ensure_future(fut, loop=loop)
[ 2221s]     
[ 2221s]             if fut.done():
[ 2221s]                 return fut.result()
[ 2221s]     
[ 2221s]             await _cancel_and_wait(fut, loop=loop)
[ 2221s]             try:
[ 2221s]                 return fut.result()
[ 2221s]             except exceptions.CancelledError as exc:
[ 2221s]                 raise exceptions.TimeoutError() from exc
[ 2221s]     
[ 2221s]         waiter = loop.create_future()
[ 2221s]         timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
[ 2221s]         cb = functools.partial(_release_waiter, waiter)
[ 2221s]     
[ 2221s]         fut = ensure_future(fut, loop=loop)
[ 2221s]         fut.add_done_callback(cb)
[ 2221s]     
[ 2221s]         try:
[ 2221s]             # wait until the future completes or the timeout
[ 2221s]             try:
[ 2221s]                 await waiter
[ 2221s]             except exceptions.CancelledError:
[ 2221s]                 if fut.done():
[ 2221s]                     return fut.result()
[ 2221s]                 else:
[ 2221s]                     fut.remove_done_callback(cb)
[ 2221s]                     # We must ensure that the task is not running
[ 2221s]                     # after wait_for() returns.
[ 2221s]                     # See https://bugs.python.org/issue32751
[ 2221s]                     await _cancel_and_wait(fut, loop=loop)
[ 2221s]                     raise
[ 2221s]     
[ 2221s]             if fut.done():
[ 2221s]                 return fut.result()
[ 2221s]             else:
[ 2221s]                 fut.remove_done_callback(cb)
[ 2221s]                 # We must ensure that the task is not running
[ 2221s]                 # after wait_for() returns.
[ 2221s]                 # See https://bugs.python.org/issue32751
[ 2221s]                 await _cancel_and_wait(fut, loop=loop)
[ 2221s]                 # In case task cancellation failed with some
[ 2221s]                 # exception, we should re-raise it
[ 2221s]                 # See https://bugs.python.org/issue40607
[ 2221s]                 try:
[ 2221s]                     return fut.result()
[ 2221s]                 except exceptions.CancelledError as exc:
[ 2221s] >                   raise exceptions.TimeoutError() from exc
[ 2221s] E                   asyncio.exceptions.TimeoutError
[ 2221s] 
[ 2221s] /usr/lib64/python3.10/asyncio/tasks.py:458: TimeoutError
[ 2221s] 
[ 2221s] During handling of the above exception, another exception occurred:
[ 2221s] 
[ 2221s] manager = <test.helpers.TestManager object at 0x7f92c80be140>
[ 2221s] 
[ 2221s]     @all_layouts_config
[ 2221s]     def test_cycle_layouts(manager):
[ 2221s]         manager.test_window("one")
[ 2221s]         manager.test_window("two")
[ 2221s]         manager.test_window("three")
[ 2221s]         manager.test_window("four")
[ 2221s]         manager.c.group.focus_by_name("three")
[ 2221s]         assert_focused(manager, "three")
[ 2221s]     
[ 2221s]         # Cycling all the layouts must keep the current window focused
[ 2221s]         initial_layout_name = manager.c.layout.info()["name"]
[ 2221s]         while True:
[ 2221s] >           manager.c.next_layout()
[ 2221s] 
[ 2221s] test/layouts/test_common.py:511: 
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] libqtile/command/client.py:199: in __call__
[ 2221s]     return self._command.execute(self._current_node, args, kwargs)
[ 2221s] libqtile/command/interface.py:232: in execute
[ 2221s]     status, result = self._client.send((call.parent.selectors, call.name, args, kwargs))
[ 2221s] libqtile/ipc.py:171: in send
[ 2221s]     return asyncio.run(self.async_send(msg))
[ 2221s] /usr/lib64/python3.10/asyncio/runners.py:44: in run
[ 2221s]     return loop.run_until_complete(main)
[ 2221s] /usr/lib64/python3.10/asyncio/base_events.py:646: in run_until_complete
[ 2221s]     return future.result()
[ 2221s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[ 2221s] 
[ 2221s] self = <libqtile.ipc.Client object at 0x7f92c8b275b0>
[ 2221s] msg = ([], 'next_layout', (), {})
[ 2221s] 
[ 2221s]     async def async_send(self, msg: Any) -> Any:
[ 2221s]         """Send the message to the server
[ 2221s]     
[ 2221s]         Connect to the server, then pack and send the message to the server,
[ 2221s]         then wait for and return the response from the server.
[ 2221s]         """
[ 2221s]         try:
[ 2221s]             reader, writer = await asyncio.wait_for(
[ 2221s]                 asyncio.open_unix_connection(path=self.socket_path), timeout=3
[ 2221s]             )
[ 2221s]         except (ConnectionRefusedError, FileNotFoundError):
[ 2221s]             raise IPCError("Could not open {}".format(self.socket_path))
[ 2221s]     
[ 2221s]         try:
[ 2221s]             send_data = _IPC.pack(msg, is_json=self.is_json)
[ 2221s]             writer.write(send_data)
[ 2221s]             writer.write_eof()
[ 2221s]     
[ 2221s]             read_data = await asyncio.wait_for(reader.read(), timeout=10)
[ 2221s]         except asyncio.TimeoutError:
[ 2221s] >           raise IPCError("Server not responding")
[ 2221s] E           libqtile.ipc.IPCError: Server not responding
[ 2221s] 
[ 2221s] libqtile/ipc.py:193: IPCError
[ 2221s] --------------------------- Captured stderr teardown ---------------------------
[ 2221s] Killing qtile forcefully
[ 2221s] qtile exited with exitcode: -9
```

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by elParaguayo** [comment]
Thanks for posting this.

Does it fail consistently on the same tests? From a quick look, I can see anything special about those tests that could cause them to fail. Most failures are in `test_manager.py` but, interestingly, not every test in that file is failing.

Is pytest running tests in parallel?

Not sure if this is related to #1634 or #2103 but they may be worth a quick look as I think they dealt with random test failures.

**Comment by Pi-Cla** [comment]
Another thing that may be relevent may be the way qtile is packaged, https://build.opensuse.org/package/view_file/home:Pi-Cla:branches:X11:windowmanagers/qtile/qtile.spec?expand=1 , in particular the %check section maybe? Note that I am not the maintainer of the qtile package, just an openSUSE user.

**Comment by elParaguayo** [comment]
You may want to look at the Arch package file (https://github.com/archlinux/svntogit-community/blob/packages/qtile/trunk/PKGBUILD) and check that all dependencies are being installed. However, most of your tests are apparently completing successfully (although I can't see the detail in the logs, only a summary saying 766 tests passed) so this doesn't feel like a dependency issue.

You're also not testing the Wayland backend but that may be a deliberate choice on your part.

**Comment by Pi-Cla** [comment]
The Wayland backend is not being tested as pywlroots and pywayland haven't been packaged yet. There is someone who is planning on packaging and sending them to the openSUSE repo, but they are waiting for openSUSE to update to the latest version of Wayland.

hmm... since we can't seem to figure out why the tests are failing, you think I should just disable them for now and send the package to the openSUSE repo? (the maintainer and others will review my submission when I send it) @elParaguayo

**Comment by uncomfyhalomacro** [comment]
Hmm now other tests are failing. :upside_down_face: i think i will have to disable some or all of the tests because i can run qtile *just fine* though I know tests are important to test to avoid any gotchas

**Comment by Pi-Cla** [comment]
Btw, @uncomfyhalomacro is also working on Qtile and has made some more progress on getting the tests to work. (though there is still some brokenness, ignore the RGB errors, they will add the RGB patches later)
https://build.opensuse.org/public/build/home:uncomfyhalomacro/openSUSE_Tumbleweed/x86_64/qtile/_log

**Comment by uncomfyhalomacro** [comment]
Here is the latest log @Pi-Cla. we have 
`[ 2224s] ===== 30 failed, 1452 passed, 23 skipped, 29 errors in 2175.01s (0:36:15) ======`
I still dont know how these 30 failed tests fail. it seems tests for those are kind of flaky because some of them gets a PASS or FAILS or ERROR, thus, those tests are inconsistent during the build. probably not on qtile's side because it seems openSUSE's modular packaging may have missed us some things or their X11 package missed some patches. idk.
[qtile_log.txt](https://github.com/qtile/qtile/files/9111235/qtile_log.txt)

**Comment by elParaguayo** [comment]
Some odd errors in that one, @uncomfyhalomacro !

Can't see why the `TreeTab` layout causes particular issues compared to other layouts.

All the migrate tests seem to fail with reference to `xcffib` not being installed but it must be as it's a core dependency of qtile!

**Comment by uncomfyhalomacro** [comment]
> `xcffib`

Going to add xcb to the test build

**Comment by uncomfyhalomacro** [comment]
It does not look promising in these though:
```
[   60s] test/test_bar.py::test_draw[1-wayland-GBConfig] PASSED                   [  0%]
[   81s] test/test_bar.py::test_prompt[1-x11-GBConfig] FAILED                     [  0%]
[   81s] test/test_bar.py::test_prompt[1-x11-GBConfig] FAILED                     [  0%]
[  103s] test/test_bar.py::test_prompt[1-wayland-GBConfig] FAILED                 [  0%]
[  103s] test/test_bar.py::test_prompt[1-wayland-GBConfig] FAILED                 [  0%]
[  103s] test/test_bar.py::test_event[1-x11-GBConfig] PASSED                      [  0%]
[  103s] test/test_bar.py::test_event[1-wayland-GBConfig] PASSED                  [  0%]
[  103s] test/test_bar.py::test_textbox[1-x11-GBConfig] PASSED                    [  0%]
[  104s] test/test_bar.py::test_textbox[1-wayland-GBConfig] PASSED                [  0%]
```

**Comment by uncomfyhalomacro** [comment]
Again. Just to add clarification during my experience, qtile works normally for daily usage (assuming people use 64-bit architecture because i586 - i386 breaks and segfaults). Unfortunately, I have to disable the tests on another build to install qtile properly for testing if it works well for daily usage. :) The only issue is what packages or possible issues that might have caused those errors to fail is a question I cannot answer. Although, I am at the point of probably disabling the tests permanently in openSUSE

**Comment by elParaguayo** [comment]
Understandable!

**Comment by jayvdb** [comment]
I may have found a source of problems.
In https://build.opensuse.org/package/show/home:jayvdb:branches:devel:languages:python/python-cairocffi I enabled the xcb tests, and they fail with
> cairocffi.CairoError: cairo returned CAIRO_STATUS_INVALID_FORMAT: b'invalid value for an input cairo_format_t'

It may not be the problem, but it certainly doesnt look good.
I dont see anything similar raised with https://github.com/Kozea/cairocffi , but it may be something else causing it.
I notice that https://build.opensuse.org/package/show/openSUSE:Factory/cairo is also out of date at 0.16 while the latest is 0.17.4

**Comment by uncomfyhalomacro** [comment]
> I may have found a source of problems. In https://build.opensuse.org/package/show/home:jayvdb:branches:devel:languages:python/python-cairocffi I enabled the xcb tests, and they fail with
> 
> > cairocffi.CairoError: cairo returned CAIRO_STATUS_INVALID_FORMAT: b'invalid value for an input cairo_format_t'
> 
> It may not be the problem, but it certainly doesnt look good. I dont see anything similar raised with https://github.com/Kozea/cairocffi , but it may be something else causing it. I notice that https://build.opensuse.org/package/show/openSUSE:Factory/cairo is also out of date at 0.16 while the latest is 0.17.4

Woah... I didn't even notice that... Anyhow, have you tried opening an SR to update cairo in opensuse?

**Comment by jayvdb** [comment]
I found a branch with 0.17 and am now testing it with cairocffi https://build.opensuse.org/package/show/home:jayvdb:cairocffi/cairo

**Comment by uncomfyhalomacro** [comment]
@jayvdb Thanks for finding another issue. I apologize for not being of any assistance. Since I have things to do at home :sweat_smile:. Hopefully that will fix it.

**Comment by jayvdb** [comment]
Unfortunately that didnt fix the cairocffi xcb tests.  I will need to keep digging.

Also when I tried running qtile `master` tests locally, my entire system froze.  Not 100% sure qtile was the cause.  However when running qtile builds on the OBS build workers, I have seen the VM halt midway through the tests.  Seems fishy.  Will dig into this more tomorrow.

**Comment by jayvdb** [comment]
And when I run the cairocffi tests locally

```py
>       surface = XCBSurface(xcb_conn, wid, root_visual, width, height)

cairocffi/test_xcb.py:181: 

self = <cairocffi.xcb.XCBSurface object at 0x7f5a7f0cea40>, conn = <xcffib.Connection object at 0x7f5a7f0d3d00>, drawable = 25165824, visual = <xcffib.xproto.VISUALTYPE object at 0x7f5a7f0d3730>, width = 10, height = 10

    def __init__(self, conn, drawable, visual, width, height):
        c_visual = visualtype_to_c_struct(visual)
    
>       p = cairo.cairo_xcb_surface_create(
            conn._conn, drawable, c_visual, width, height)
E       AttributeError: cffi library 'libcairo.so.2' has no function, constant or global variable named 'cairo_xcb_surface_create'

cairocffi/xcb.py:37: AttributeError
```

Ubuntu & Debian (and many others) are still on cairo 0.16 https://repology.org/project/cairo/versions , so that being outdated probably isnt important.

**Comment by uncomfyhalomacro** [comment]
> Unfortunately that didnt fix the cairocffi xcb tests.  I will need to keep digging.
> 
> 
> 
> Also when I tried running qtile `master` tests locally, my entire system froze.  Not 100% sure qtile was the cause.  However when running qtile builds on the OBS build workers, I have seen the VM halt midway through the tests.  Seems fishy.  Will dig into this more tomorrow.

the halt? u mean the builds froze? i experienced that before and it happens randomly, other builds don't halt (but fail) while others did halt.

**Comment by jayvdb** [comment]
I have got cairocffi xcb tests operational on OBS, but I get a seg fault on some of the builds.
https://build.opensuse.org/package/show/home:jayvdb:branches:devel:languages:python/python-cairocffi
```
[   37s] + xvfb-run '--server-args=-screen 0 1280x1024x16' python3.10 -m pytest
[   42s] ============================= test session starts ==============================
[   42s] platform linux -- Python 3.10.5, pytest-7.1.1, pluggy-1.0.0
[   42s] rootdir: /home/abuild/rpmbuild/BUILD/cairocffi-1.3.0, configfile: setup.cfg
[   42s] collected 54 items
[   42s] 
[   42s] test_cairo.py ..............................................             [ 85%]
[   42s] test_numpy.py .                                                          [ 87%]
[   42s] test_pixbuf.py .x...                                                     [ 96%]
[   42s] test_xcb.py .Fatal Python error: Segmentation fault
[   42s] 
[   42s] Current thread 0x00007fff828d3b20 (most recent call first):
[   42s]   File "/home/abuild/rpmbuild/BUILDROOT/python-cairocffi-1.3.0-53.1.ppc64le/usr/lib/python3.10/site-packages/cairocffi/xcb.py", line 37 in __init__
[   42s]   File "/home/abuild/rpmbuild/BUILD/cairocffi-1.3.0/tests/test_xcb.py", line 181 in test_xcb_window
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/python.py", line 1761 in runtest
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 166 in pytest_runtest_call
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 259 in <lambda>
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 338 in from_call
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 258 in call_runtest_hook
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 219 in call_and_report
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 130 in runtestprotocol
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/main.py", line 347 in pytest_runtestloop
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/main.py", line 322 in _main
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/main.py", line 268 in wrap_session
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
[   42s]   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/config/__init__.py", line 164 in main
[   42s]   File "/usr/lib/python3.10/site-packages/_pytest/config/__init__.py", line 187 in console_main
[   42s]   File "/usr/lib/python3.10/site-packages/pytest/__main__.py", line 5 in <module>
[   42s]   File "/usr/lib64/python3.10/runpy.py", line 86 in _run_code
[   42s]   File "/usr/lib64/python3.10/runpy.py", line 196 in _run_module_as_main
[   42s] 
[   42s] Extension modules: _cffi_backend, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg.lapack_lite, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator (total: 15)
[   42s] /usr/bin/xvfb-run: line 168:  1729 Segmentation fault      DISPLAY=:$SERVERNUM XAUTHORITY=$AUTHFILE "$@" 2>&1
[   42s] error: Bad exit status from /var/tmp/rpm-tmp.CFtFmG (%check)
```

**Comment by jayvdb** [comment]
This problem still exists, somewhere, but I have updated the openSUSE package in Factory, by using `pytest -k 'not (TreeTab or fakescreen or manager or test_window or test_prompt or test_cycle_layouts or test_multiple_borders or (notify and test_parse_text))'`

**Comment by uncomfyhalomacro** [comment]
Bummer. I guess the failing tests is probably the reason we have to disable in the future if this keeps up. It seems most of the failing tests come from cairocffi and xcb. And yet we haven't found anything conclusive for what caused the tests to segfault or fail. One thing I am suspicious of is `DISPLAY=:$SERVERNUM XAUTHORITY=$AUTHFILE "$@" 2>&1` from the segfaults.
